src/java/org/apache/cassandra/replication/CoordinatedTransfer.java

```diff
@@ -0,0 +1,586 @@
+
+    final ConcurrentMap<InetAddressAndPort, SingleTransferResult> streams;
+
+    final Collection<SSTableReader> sstables;
+
+    final ConsistencyLevel cl;
+
+    final Supplier<MutationId> getActivationId;
+    volatile MutationId activationId = null;
+
+    CoordinatedTransfer(String keyspace, Range<Token> range, Participants participants, Collection<SSTableReader> sstables, ConsistencyLevel cl, Supplier<MutationId> getActivationId)
+    {
+        this(TimeUUID.Generator.nextTimeUUID(), keyspace, range, participants, sstables, cl, getActivationId);
+    }
+
+    @VisibleForTesting
+    CoordinatedTransfer(TimeUUID transferId, Range<Token> range, MutationId activationId)
+    {
+        this.transferId = transferId;
+        this.keyspace = null;
+        this.range = range;
+        this.sstables = Collections.emptyList();
+        this.cl = null;
+        this.getActivationId = () -> activationId;
+        this.activationId = activationId;
+        this.streams = new ConcurrentHashMap<>();
+    }
+
+    private CoordinatedTransfer(TimeUUID transferId, String keyspace, Range<Token> range, Participants participants, Collection<SSTableReader> sstables, ConsistencyLevel cl, Supplier<MutationId> getActivationId)
+    {
+        this.transferId = transferId;
+        this.keyspace = keyspace;
+        this.range = range;
+        this.sstables = sstables;
+        this.cl = cl;
+        this.getActivationId = getActivationId;
+
+        ClusterMetadata cm = ClusterMetadata.current();
+        this.streams = new ConcurrentHashMap<>(participants.size());
+        for (int i = 0; i < participants.size(); i++)
+        {
+            InetAddressAndPort addr = cm.directory.getNodeAddresses(new NodeId(participants.get(i))).broadcastAddress;
+            this.streams.put(addr, SingleTransferResult.Init());
+        }
+    }
+
+    void execute()
+    {
+        logger.debug("Executing tracked bulk transfer {}", this);
+        LocalTransfers.instance().save(this);
+        stream();
+    }
+
+    private void stream()
+    {
+        // TODO: Don't stream multiple copies over the WAN, send one copy and indicate forwarding
+        List<Future<Void>> streaming = new ArrayList<>(streams.size());
+        for (InetAddressAndPort to : streams.keySet())
+        {
+            Future<Void> stream = LocalTransfers.instance().executor.submit(() -> {
+                stream(to);
+                return null;
+            });
+            streaming.add(stream);
+        }
+
+        Future<List<Void>> future = FutureCombiner.allOf(streaming);
+        try
+        {
+            future.get();
+        }
+        catch (InterruptedException | ExecutionException e)
+        {
+            logger.error("{} Failed due to", logPrefix(), e);
+            Throwable cause = e instanceof ExecutionException ? e.getCause() : e;
+            throw cause instanceof RuntimeException ? (RuntimeException) cause : new RuntimeException(cause);
```

Comment:
```suggestion
            throw Throwables.unchecked(cause);
```
Happens in a few other places below...